### PR TITLE
The new loopback network interface

### PR DESCRIPTION
**Pull request by htibosch**
Description
-----------
In the earlier IPv4-only library, there has been a loopback functionality. It needed adaptation to the network interface: outgoing packets for 127.0.0.0 would be passed back to the IP-task.

The current implementation will add a new Loopback interface. It will make sure that all "outgoing" packets will be bounced back to the IP-task. The interface will get two endpoint: one for IPv4 and one for IPv6.

Test Steps
-----------
I created a demo that starts at most 4 tasks: UDPv4, UDPv6, TCPv4, and TCPv6. Each task opens two sockets that will connect and exchanges data packets.

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------
As for now, the driver needs the changes from [PR #1020](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/1020).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by rawalexe** [comment]
Greetings,
Thank you so much for your contribution, I'll pass this along to the team and have some one review it/

File: source/portable/NetworkInterface/loopback/loopbackNetworkInterface.c [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                if( xIsIPv6Loopback( &( pxDescriptor->xIPAddress ) ) != pdFALSE )
+                {
+                    vNDRefreshCacheEntry( &xMACAddress, &( pxDescriptor->xIPAddress.xIP_IPv6 ), pxDescriptor->pxEndPoint );
+                }
+            #endif
+        }
+        else
+        {
+            #if ( ipconfigUSE_IPv4 != 0 )
+                if( xIsIPv4Loopback( pxDescriptor->xIPAddress.ulIP_IPv4 ) )
```
Comment (tony-josi-aws) [inline source/portable/NetworkInterface/loopback/loopbackNetworkInterface.c:151]: I couldn't find a definition for `xIsIPv4Loopback`, as we didn't had that function defined in the stack is it missed?

Comment (tony-josi-aws) [reply source/portable/NetworkInterface/loopback/loopbackNetworkInterface.c:151]: Marking this as resolved as #1020 adds the mentioned change.

**Review by shubnil — COMMENTED** [review]
Thanks Hein. This is a good PR. I have now started thinking what is a good way to implement a loopback:
1. Software (Network Interface Loopback) as in this PR
2. Hardware Loopback (MAC or PHY loopback) - All MAC and PHY loopback registers have standard offsets from base address and hence this code will also be generic for all Hardwares.
3. Have Both the above and give the configuration option to the user.

**Review by tony-josi-aws — APPROVED** [review]

**Comment by shubnil** [comment]
I think we can merge the PR as a first step and create an internal ticket for adding Hardware loopback as a next step enhancement.

**Review by shubnil — APPROVED** [review]

**Comment by htibosch** [comment]
> create an internal ticket for adding Hardware loopback

I think that "Hardware loopback" is a modus of the EMAC in which the outgoing packets are coupled back to the incoming stream. I used it sometimes as a testing mode.

The loopback network interface that I created works without hardware. The loopback is realised by sending [`eNetworkRxEvent` messages](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/main/source/portable/NetworkInterface/loopback/loopbackNetworkInterface.c#L170) to the IP-stack.

I do not see a way in which a hardware loopback can help us realise a loopback device.

**Pull request merged**